src/hydration/hydration.ts

```diff
@@ -0,0 +1,155 @@
+import { DEFAULT_STALE_TIME, DEFAULT_CACHE_TIME } from '../core/config'
+
+import type { Query, QueryCache, QueryKey, QueryConfig } from 'react-query'
+
+export interface DehydratedQueryConfig<TResult> {
+  staleTime?: number
+  cacheTime?: number
+  initialData?: TResult
+}
+
+export interface DehydratedQuery<TResult> {
+  config: DehydratedQueryConfig<TResult>
+  updatedAt: number
+}
+
+export interface DehydratedQueries<TResult> {
+  [hash: string]: DehydratedQuery<TResult>
+}
+
+export type QueryKeyParserFunction = (queryHash: string) => QueryKey
+export type ShouldHydrateFunction = <TResult>({
+  queryKey,
+  updatedAt,
+  staleTime,
+  cacheTime,
+  data,
+}: {
+  queryKey: QueryKey
+  updatedAt: number
+  staleTime: number
+  cacheTime: number
+  data?: TResult
+}) => boolean
+export interface HydrateConfig {
+  queryKeyParserFn?: QueryKeyParserFunction
+  shouldHydrate?: ShouldHydrateFunction
+  activateTimeoutsManually?: boolean
+}
+
+export type ShouldDehydrateFunction = <TResult, TError = unknown>(
+  query: Query<TResult, TError>
+) => boolean
+export interface DehydrateConfig {
+  shouldDehydrate?: ShouldDehydrateFunction
+}
+
+function dehydrateQuery<TResult, TError = unknown>(
+  query: Query<TResult, TError>
+): DehydratedQuery<TResult> {
+  const dehydratedQuery: DehydratedQuery<TResult> = {
+    config: {},
+    updatedAt: query.state.updatedAt,
+  }
+
+  // Most config is not dehydrated but instead meant to configure again when
+  // consuming the de/rehydrated data, typically with useQuery on the client.
+  // Sometimes it might make sense to prefetch data on the server and include
+  // in the html-payload, but not consume it on the initial render.
+  // We still schedule stale and garbage collection right away, which means
+  // we need to specifically include staleTime and cacheTime in dehydration.
+  if (query.config.staleTime !== DEFAULT_STALE_TIME) {
+    dehydratedQuery.config.staleTime = query.config.staleTime
+  }
+  if (query.config.cacheTime !== DEFAULT_CACHE_TIME) {
+    dehydratedQuery.config.cacheTime = query.config.cacheTime
+  }
+  if (query.state.data !== undefined) {
+    dehydratedQuery.config.initialData = query.state.data
+  }
+
+  return dehydratedQuery
+}
+
+export function dehydrate<TResult>(
+  queryCache: QueryCache,
+  dehydrateConfig?: DehydrateConfig
+): DehydratedQueries<TResult> {
+  const config = dehydrateConfig || {}
+  const { shouldDehydrate } = config
+  const dehydratedQueries: DehydratedQueries<TResult> = {}
+  for (const [queryHash, query] of Object.entries(queryCache.queries)) {
+    if (query.state.status === 'success') {
+      if (shouldDehydrate && !shouldDehydrate(query)) {
+        continue
+      }
+      dehydratedQueries[queryHash] = dehydrateQuery(query)
+    }
+  }
+
+  return dehydratedQueries
+}
+
+export function hydrate<TResult>(
```

Comment:
Not sure if this is even particularly helpful/desireable (and I'm no TS expert either) but I think you *could* make `hydrate` only return the `activateTimeouts` function when `activateTimeoutsManually=true`. I think the types to support it would be something like this using constant overloads:

```tsx
function hydrate<TResult>(
  queryCache: QueryCache,
  dehydratedQueries: unknown;
  hydrateConfig?: HydrateConfig;
): void | () => void;

function hydrate<TResult>(
  queryCache: QueryCache,
  dehydratedQueries: unknown;
  hydrateConfig: Omit<HydrateConfig, "activateTimeoutsManually"> & { activateTimeoutsManually: true };
): () => void;

function hydrate<TResult>(
  queryCache: QueryCache,
  dehydratedQueries: unknown;
  hydrateConfig: Omit<HydrateConfig, "activateTimeoutsManually"> & { activateTimeoutsManually?: false };
): void;
```

https://stackoverflow.com/questions/25671381/can-i-overload-function-on-constant-of-number-type

Reply:
Nice! I thought it was probably possible but didn't have the TS-skills for it myself. 😄  I'm not sure what's desirable either, but I also thought of this kind of like "whatever the config, if you call the function returned by hydrate you can be sure that timeouts have been activated".

If you change the config option from manual to automatic and forget to remove the `activateTimeouts()` call you'll get an error with your version, which is good because you find a now pretty useless call to `activateTimeouts()`, but bad if this happens at runtime in prod because you are not using TS. 😅  Should be unlikely though, I don't have strong feelings on this either way really. 🤷